Add IODA converter for ARO observations

ibanos90 opened the pull request:
## Description
This PR incorporates the IODA converter for airborne GNSS radio occultation (ARO) observations. It follows very closely the existing python converter for spaceborne GNSS RO with the necessary modifications for ARO NETCDF files. A ctest was added to ensure the converter works correctly.

In addition to the variables already existing in the spaceborne GNSS RO converter, the geopotential height calculated in the converter and retrieved partial bending angle were added in the MetaData group. The geopotential height is used in the ARO ROPP2D operator to obtain the radius of the aircraft which is limit of the integration. On the other hand, the retrieved partial bending angle is added for future use.

## Acceptance Criteria
Passes the ctest.

hailingz commented:
Ctests failed and I just took a quick look about that.
You uploaded a file "test/testoutput/2021120600_atms_sdr.nc4". This file is used for ATMS radiance converter ctest.
Using this name caused the failure of "test_iodaconv_atms_nc4". 
If you want to add your own test (and you need to complete this PR), you can follow the test_iodaconv_atms_nc4 example as listed by the below codes.
https://github.com/JCSDA-internal/ioda-converters/blob/develop/test/CMakeLists.txt#L1578
Also I suggest you change your testinput file name to reflect the content of the file.  

In summary, the steps are run your converter using your input file, upload your input and output files to testinput and testoutput respectively if you add a ctest in test/CMakeLists.txt
Let me or @srherbener know if you need further help.

BenjaminRuston commented:
thank you @hailingz ,, yes @ibanos90 the new output you should test against is your own which you produce and will be a new file in IODA format

the term `atms` stands for `advanced technology microwave sensor` so you use whatever name you like `aro_atmPrf` or something like this for a preface should work, and note please do not use the same name as the input file.

ibanos90 commented:
Thanks @hailingz and @BenjaminRuston for your comments. It seems that the file "test/testoutput/2021120600_atms_sdr.nc4" got here when doing a merge with develop or something. It is not part of the ARO converter. I will make a commit removing that file. Also, I do not use 'atms' in the converter, input or output names. 

Thanks @hailingz for suggesting changing the name of the testinput file. Indeed it can be more explicit of the file content.
Also, I will add a ctest. However, I will need some help to make the IODA writer to work, which is still not working as I mentioned in the description of this draft PR. Thanks!

ibanos90 commented:
Hi all, the converter is working properly now, writing the ioda output file. I added the output file as well as the ctest and it passes successfully. Please let me know if you get the same results.

hailingz commented:
The following ctests are still failing. It will be helpful if you could run those ctests on your cluster to see if they pass. If all tests pass on your cluster but fail on AWS please let me know. Thanks!




The following tests FAILED:
--
952 | 61 - test_iodaconv_gnssaro_netcdf_conv (Failed)
953 | 108 - test_iodaconv_atms_nc4 (Failed)

ibanos90 commented:
> The following ctests are still failing. It will be helpful if you could run those ctests on your cluster to see if they pass. If all tests pass on your cluster but fail on AWS please let me know. Thanks!
> 
> ## The following tests FAILED:
> 952 | 61 - test_iodaconv_gnssaro_netcdf_conv (Failed) 
> 953 | 108 - test_iodaconv_atms_nc4 (Failed)

Sure, I will do that. Thanks @hailingz!

ibanos90 commented:
@hailingz after adding back the atms testoutput file, the only ctest that fails is test_iodaconv_sondetac.
```
    Start 603: test_iodaconv_sondetac
1/2 Test #603: test_iodaconv_sondetac ...........***Failed    1.50 sec
Traceback (most recent call last):
  File "/glade/p/mmm/parc/ivette/sio/codes-builds/ioda-bundle_aro/build4/bin/sonde_tac2ioda.py", line 23, in <module>
    from cartopy import geodesic
ModuleNotFoundError: No module named 'cartopy'
```

hailingz commented:
I added the python PATH and now all tests pass.
You have four variables in the "ObsValue" group. While I am not sure if "partialBendingAngle" is a observable (which you will assimilate), I believe "geopotentialHeight" does not belong to the ObsValue group.

ibanos90 commented:
> I added the python PATH and now all tests pass. You have four variables in the "ObsValue" group. While I am not sure if "partialBendingAngle" is a observable (which you will assimilate), I believe "geopotentialHeight" does not belong to the ObsValue group.

Hi @hailingz, thanks so much for your help on this! In my last commit I moved the geopotential height to the MetaData group as well as the partial bending angle. However, in the future we could discuss if partial bending angle could be moved to the ObsValue group. Currently we do not use the partial bending angle but we could use it in the future.

hailingz commented:
please also update your testoutput file as you changed your converter.

ibanos90 commented:
> please also update your testoutput file as you changed your converter.

That's right. Thanks for reminding me!

hailingz commented:
I think this PR is ready to review! @srherbener @BenjaminRuston

ibanos90 commented:
> I think this PR is ready to review! @srherbener @BenjaminRuston

Thanks @hailingz! Should I close this one and open a PR?

srherbener commented on src/gnssro/gnssaro_netcdf2ioda.py:
```diff
@@ -0,0 +1,494 @@
+    return meta_data_types
+
+
+def concat_obs_dict(obs_data, append_obs_data):
```
This looks like a copy of `concat_obs_dict` in `src/lib-python/def_jedi_utils.py `with the addition of a few more data types when appending with fill values. If I've got this right, please consider updating `concat_obs_dict` in `src/lib/lib-python/def_jedi_utils.py` and using that version. This would help with sharing common code. I think there are other scripts using `concat_obs_dict`.

ibanos90 replied:
Hi @srherbener, thanks for your comment! You are right, I took a look and the functions are very similar, I didn't realize that before, just used what was already in `gnssro_bufr2ioda.py`. I will update `concat_obs_dict` in `src/lib-python/def_jedi_utils.py` in a commit. I will see how that changes for other scripts that function.

BenjaminRuston replied:
@ibanos90 I did not get to this

PatNichols reviewed (commented):
This needs to be updated to the new common calling structure. It's only few lines of code.

PatNichols commented on src/gnssro/gnssaro_netcdf2ioda.py:
```diff
@@ -0,0 +1,494 @@
+import netCDF4 as nc
+import h5py as h5
+
+IODA_CONV_PATH = Path(__file__).parent/"@SCRIPT_LIB_PATH@"
```
This is the place where an update is needed.

BenjaminRuston reviewed (approved):
tested after making changes to new calling structure and seems to work as expected

BenjaminRuston commented:
@ibanos90 this is working for me please test, not you will have to now do an additional append of the PYTHONPATH, or you can do a pip install.

```PYTHONPATH="${path_of_ioda_bundle}/ioda-bundle/iodaconv/src:${PYTHONPATH}" ```

or (assuming in a virtual environment)

```cd ${path_of_ioda_bundle}/ioda-bundle/iodaconv/src
python3 -m pip install -e .```

ibanos90 commented:
> @ibanos90 this is working for me please test, *note you will have to now do an additional append of the PYTHONPATH, or you can do a pip install.
> 
> `PYTHONPATH="${path_of_ioda_bundle}/ioda-bundle/iodaconv/src:${PYTHONPATH}" `
> 
> or (assuming in a virtual environment)
> 
> ```
> python3 -m pip install -e .
> ```

Thanks for the fix @BenjaminRuston! I will it now.

srherbener reviewed (approved):
Thanks for addressing my comments!

ibanos90 commented:
Thank you all for your reviews and help!
